File: backend-ruby/app/controllers/api/v1/diem_danh_controller.rb
class Api::V1::DiemDanhController < ApplicationController
  before_action :set_diem_danh, only: [:show, :update]

  VALID_STATUSES = %w[present absent late excused].freeze

  def index
    records = DiemDanh
              .includes(:dang_ky_thi, :ky_thi, :phong_thi)
              .order(DiemDanhID: :asc)

    render json: records.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        phong_thi: {
          only: [:PhongThiID, :MaPhong, :TenPhong]
        },
        dang_ky_thi: {
          only: [:DangKyThiID, :SoBaoDanh],
          include: {
            sinh_vien: {
              only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
            }
          }
        }
      }
    )
  end

  def show
    render json: @diem_danh.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        phong_thi: {
          only: [:PhongThiID, :MaPhong, :TenPhong]
        },
        dang_ky_thi: {
          only: [:DangKyThiID, :SoBaoDanh],
          include: {
            sinh_vien: {
              only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
            }
          }
        }
      }
    )
  end

  def create
    phan_phong = PhanPhong.find_by!(DangKyThiID: diem_danh_params[:DangKyThiID])

    status = diem_danh_params[:TrangThai] || "absent"

    unless VALID_STATUSES.include?(status)
      return render json: { error: "Invalid attendance status." }, status: :unprocessable_entity
    end

    record = DiemDanh.new(diem_danh_params)
    record.KyThiID = phan_phong.KyThiID
    record.PhongThiID = phan_phong.PhongThiID
    record.TrangThai = status
    record.ThoiGianCheckIn ||= Time.current if %w[present late].include?(status)

    if record.save
      render json: record, status: :created
    else
      render json: { errors: record.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotFound
    render json: { error: "Room assignment not found for this registration." }, status: :not_found
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Attendance already exists for this registration." }, status: :conflict
  rescue ActiveRecord::InvalidForeignKey => e
    render json: { error: e.message }, status: :conflict
  end

  def update
    status = diem_danh_params[:TrangThai]

    if status.present? && !VALID_STATUSES.include?(status)
      return render json: { error: "Invalid attendance status." }, status: :unprocessable_entity
    end

    attrs = diem_danh_params.to_h

    if %w[present late].include?(status) && @diem_danh.ThoiGianCheckIn.blank?
      attrs[:ThoiGianCheckIn] = Time.current
    end

    if %w[absent excused].include?(status)
      attrs[:ThoiGianCheckIn] = nil
    end

    if @diem_danh.update(attrs)
      render json: @diem_danh
    else
      render json: { errors: @diem_danh.errors.full_messages }, status: :unprocessable_entity
    end
  end

  private

  def set_diem_danh
    @diem_danh = DiemDanh.find(params[:id])
  end

  def diem_danh_params
    params.require(:diem_danh).permit(
      :DangKyThiID,
      :TrangThai,
      :ThoiGianCheckIn,
      :NguoiGhiNhanID,
      :GhiChu
    )
  end
end

File: backend-csharp/Controllers/DiemDanhController.cs
using System.Text.Json;
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/diem_danh")]
public class DiemDanhController(ExamDbContext dbContext) : ApiControllerBase
{
    private static readonly HashSet<string> ValidStatuses = ["present", "absent", "late", "excused"];

    [HttpGet]
    public async Task<IActionResult> Index()
    {
        List<DiemDanh> records = await QueryWithIncludes().OrderBy(record => record.DiemDanhID).ToListAsync();
        return Ok(records);
    }

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        DiemDanh? record = await QueryWithIncludes().FirstOrDefaultAsync(item => item.DiemDanhID == id);
        return record is null ? NotFound(Error("Attendance record not found.")) : Ok(record);
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] JsonElement body)
    {
        DiemDanh? record = ReadBody<DiemDanh>(body, "diem_danh");
        if (record is null)
        {
            return UnprocessableEntity(Errors("Invalid attendance payload."));
        }

        PhanPhong? phanPhong = await dbContext.PhanPhong.FirstOrDefaultAsync(item => item.DangKyThiID == record.DangKyThiID);
        if (phanPhong is null)
        {
            return NotFound(Error("Room assignment not found for this registration."));
        }

        string status = string.IsNullOrWhiteSpace(record.TrangThai) ? "absent" : record.TrangThai;
        if (!ValidStatuses.Contains(status))
        {
            return UnprocessableEntity(Error("Invalid attendance status."));
        }

        record.KyThiID = phanPhong.KyThiID;
        record.PhongThiID = phanPhong.PhongThiID;
        record.TrangThai = status;
        record.ThoiGianCheckIn ??= status is "present" or "late" ? DateTime.Now : null;

        dbContext.DiemDanh.Add(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return CreatedAtAction(nameof(Show), new { id = record.DiemDanhID }, record);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Attendance already exists for this registration."));
        }
    }

    [HttpPatch("{id:int}")]
    [HttpPut("{id:int}")]
    public async Task<IActionResult> Update(int id, [FromBody] JsonElement body)
    {
        DiemDanh? record = await dbContext.DiemDanh.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error("Attendance record not found."));
        }

        DiemDanh? payload = ReadBody<DiemDanh>(body, "diem_danh");
        if (payload is null)
        {
            return UnprocessableEntity(Errors("Invalid attendance payload."));
        }

        if (!string.IsNullOrWhiteSpace(payload.TrangThai) && !ValidStatuses.Contains(payload.TrangThai))
        {
            return UnprocessableEntity(Error("Invalid attendance status."));
        }

        record.DangKyThiID = payload.DangKyThiID == 0 ? record.DangKyThiID : payload.DangKyThiID;
        record.TrangThai = string.IsNullOrWhiteSpace(payload.TrangThai) ? record.TrangThai : payload.TrangThai;
        record.NguoiGhiNhanID = payload.NguoiGhiNhanID;
        record.GhiChu = payload.GhiChu;

        if (record.TrangThai is "present" or "late" && record.ThoiGianCheckIn is null)
        {
            record.ThoiGianCheckIn = DateTime.Now;
        }

        if (record.TrangThai is "absent" or "excused")
        {
            record.ThoiGianCheckIn = null;
        }

        if (payload.ThoiGianCheckIn.HasValue)
        {
            record.ThoiGianCheckIn = payload.ThoiGianCheckIn;
        }

        record.CapNhatLuc = DateTime.Now;
        await dbContext.SaveChangesAsync();
        return Ok(record);
    }

    private IQueryable<DiemDanh> QueryWithIncludes()
    {
        return dbContext.DiemDanh
            .Include(record => record.KyThi)
            .Include(record => record.PhongThi)
            .Include(record => record.DangKyThi)
                .ThenInclude(record => record!.SinhVien);
    }
}
